Pre aux UO grouping update

jthano opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
This PR is meant to address the potential issue described by #18189 and should closes #18189.

## Design
Previously, if a UO was found to be listed as a dependency for any auxkernel, then that UO was added to the `PRE_AUX` group regardless of what the execution flag that the auxkernel was set to. Now, the flag for the auxkernel is accounted for when determining UO dependency grouping.

## Impact
No tests needed to be changed related to this code change. It is possible that this code change may affect some existing user input, and slight input change may be required to get the same result. This would occur if an auxkernel depended on a UO, but the exec flag was `FINAL`. As described in #18189, the UO in question would be executed in the `PRE_AUX` group always, even though this dependency is not really appropriate based on the input. With this code update, the UO would only be grouped in `PRE_AUX` for exec flag `FINAL`. The old behavior can be achieved by adding `force_preaux = true` to the UO input block.


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

Commit:
Updates to change the way UO are grouped into preaux group

Commit:
test added to test code changes made to address #18189

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/773823/) on 456ef53 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18293/site)

This comment will be updated on new commits.

YaqiWang commented on test/tests/userobjects/pre_aux_based_on_exec_flag/pre_post_aux_test.i:
```diff
@@ -0,0 +1,113 @@
+  # the most up to date information. scale2 is pushed into PRE_AUX and hence picks
+  # up the value of total_u from the last timestep.
+  #
+  # Additionally, the aux kernel NormalizationAuxTest2 uses scale1. If the
```
NormalizationAuxTest2 => NormalizationAuxTest1

YaqiWang commented on test/tests/userobjects/pre_aux_based_on_exec_flag/tests:
```diff
@@ -0,0 +1,22 @@
+    csvdiff = 'pre_post_aux_test1_out.csv'
+    cli_args = 'Outputs/file_base=pre_post_aux_test1_out'
+
+    requirement = 'When considering that an auxkernel depends on a UO, the exec flag will be taken into consideration to determin when the UO must be executed preaux.'
```
This requirement does not follow MOOSE standard like `MOOSE shall...`.

YaqiWang commented on test/tests/userobjects/pre_aux_based_on_exec_flag/pre_post_aux_test.i:
```diff
@@ -0,0 +1,113 @@
+  # for, the scale1 would be forced into PRE_AUX by the dependency and
+  # scale1 and scale2 would behave the same. When the dependency is handled
+  # correctly, this does not occur
+  # When NormalizationAuxTest3 is active, all posprocessors then have
```
Add: User objects including postprocessors and their depending user objects are automatically put into PRE_AUX group when the user objects are used by an auxiliary kernel on the same execute_on flag. Thus when ...

YaqiWang reviewed (changes requested):
The changes in Attributes requires @rwcarlsen 's review. You will also need to fix the test failures.

YaqiWang commented on test/tests/userobjects/pre_aux_based_on_exec_flag/tests:
```diff
@@ -0,0 +1,22 @@
+[Tests]
+  issues = '#18189'
+  design = '/GeneralUserObject.md'
```
You will need to improve a markdown file to explain the dependency a little more. I remember we have a markdown file explaining how a uo can get into pre_aux group. You will need to find it.

YaqiWang commented:
The failures in Bison and in modules possibly mean they were abusing the old behavior. You will need to get help from their developers. Possibly you will need to look into the tests a little so that you can explain to them what has happened.

rwcarlsen commented on framework/src/base/Attributes.C:
```diff
@@ -309,21 +309,38 @@ AttribPreIC::isEqual(const Attribute & other) const
 {
   auto a = dynamic_cast<const AttribPreAux *>(&other);
-  return a && (a->_val == _val);
+  if (!a || a->_vals.size() < 1)
```
Why should this return false if _vals.size is zero?  If _vals.size() and other._vals.size() are both zero - then shouldn't it return true?  Or if you expect _vals to never be size zero than you should assert that here.

jthano replied:
Firstly, I'll note this new implementation of `AttribPreAux` is based entirely on the existing attribute `AttribExecOns`. They are now very similar classes.

If _vals.size() is zero, this function returns false because it means that there are no exec flags in `a` where `AttribPreAux` should return true so this is a short cut if statement as is present in `AttribExecOns::isMatch`.

Would it always be a programming error for `a` to have a _vals vector with `size()!=1`? I'm not sure about that. If it is an error, perhaps we should change 'AttribExecOns::isMatch' as well

rwcarlsen replied:
I think both being zero length either needs to be asserted against or resolve to returning true here.

rwcarlsen commented on framework/src/base/Attributes.C:
```diff
@@ -309,21 +309,38 @@ AttribPreIC::isEqual(const Attribute & other) const
-  return a && (a->_val == _val);
+  if (!a || a->_vals.size() < 1)
+    return false;
+  auto cond = a->_vals[0];
+  if (cond == Moose::ALL)
+    return true;
+
+  for (auto val : _vals)
+    if (val == Moose::ALL || val == cond)
+      return true;
+  return false;
```
You've allowed _vals to be a vector, yet you aren't comparing all entries to all entries.  Either a vector is the wrong data structure, or you need to account for this.

jthano replied:
I'll note this new implementation of `AttribPreAux` is based entirely on the existing attribute `AttribExecOns`. They are now very similar classes. I am willing to change the design of this class, but it will then be different from `AttribExecOns` even though `AttribPreAux` and `AttribExecOns` have nearly the same functionality.

The only difference between `AttribExecOns` and `AttribPreAux`  now is that `AttribExecOns` can fill its `_vals` object by a call to `AttribExecOns::initFrom(const MooseObject * obj)`, but that won't work for `AttribPreAux` so instead, an additional constructor is added which can be used to fill `_vals` 

The reason that `_vals` is a vector is so that there only needs to be one data member. The attribute classes appear to be used in two ways, one way to store attributes of an object and in a different way when querying for objects. `_vals` is filled with more than one value only when calling the constructor that takes a vector.  This constructor is used only when calling `TheWarehouse.update()` method and the vector is filled with all exec flags which are relevant to the object passed to `update()`. The `_vals` vector is filled with only single value when calling `TheWarehouse.query` like in this block of code:
```c++
void
FEProblemBase::computeUserObjects(const ExecFlagType & type, const Moose::AuxGroup & group)
{
  TheWarehouse::Query query =
      theWarehouse().query().condition<AttribSystem>("UserObject").condition<AttribExecOns>(type);
  computeUserObjectsInternal(type, group, query);
}

void
FEProblemBase::computeUserObjectsInternal(const ExecFlagType & type,
                                          const Moose::AuxGroup & group,
                                          TheWarehouse::Query & query)
{
  if (group == Moose::PRE_IC)
    query.condition<AttribPreIC>(true);
  else if (group == Moose::PRE_AUX)
    query.condition<AttribPreAux>(type);
```

rwcarlsen commented on framework/src/base/Attributes.C:
```diff
@@ -309,21 +309,38 @@ AttribPreIC::isEqual(const Attribute & other) const
+  if (!a || a->_vals.size() != _vals.size())
+    return false;
+
+  for (size_t i = 0; i < a->_vals.size(); i++)
+    if (a->_vals[i] != _vals[i])
+      return false;
```
What if they have the same entries, but are in a different order?

jthano replied:
> But this code promotes a UO to pre aux for _all_ execute on flags if an aux kernel running on only _one_ execute on flag depends on that uo. Is this what you wanted? Or do we need to revisit the implementation to try to achieve the full handling I just described?

I don't think this is the case, the test I added I think shows this does not occur. In the test I added, the section of input shown below is the important part. Before the code changes proposed in this MR, the `scale1` UO is forced into the `PRE_AUX` group because `NormalizationAuxTest1` depends on `scale1`. But `NormalizationAuxTest1` only execs on `FINAL` so there is not a dependency for any other exec flag. When I run this test with the code in this MR, I see that `scale1` is not run in `PRE_AUX` except for `FINAL`

```
[AuxKernels]
  active = 'NormalizationAuxTest1'
  [NormalizationAuxTest1]
    type = NormalizationAux
    variable = w
    source_variable = u
    normalization = scale1
    execute_on = 'FINAL'
  []
  ...
  [Postprocessors]
  ...
  [./scale1]
    type = ScalePostprocessor
    value = total_u
    scaling_factor = 1
  [../]
```

jthano replied:
I copied this logic from `AttribExecOns::isEqual(const Attribute & other) const` but I'm not sure why `isEqual` needs to check for identical ordering. Do you now why `AttribExecOns` checks for ordering? If there is no reason, then I think changing this to not check the ordering makes sense.

jthano replied:
I haven't changed this in my most recent code updates. I still want `AttribPreAux` and `AttribPostAux` to be similar to `AttribExecOns` which uses a `_vals` a vector and includes this logic to check `isEqual`. However, if you think I should change `_vals` to a std::set for both `AttribPreAux` and `AttribPostAux`, I can change the logic of `AttribPreAux::isEqual` and  `AttribPostAux::isEqual` to check if the sets are the same

rwcarlsen replied:
Yes - this needs to be order independent.  AttribExecOns is a bit different in that its vector is only ever populated internally inside its initFrom function - where it retrieves the exec on flags from an ordered source.  In the code here, however, the pre/post aux flags are inserted externally - i.e. passed in via a constructor arg - so the only way they are ordered is if the calling code knows to only pass in an ordered vector - which is much less safe.  So this needs to be sorted.

rwcarlsen commented on framework/include/base/Attributes.h:
```diff
@@ -285,19 +285,30 @@ class AttribPreIC : public Attribute
 
 private:
-  bool _val = false;
+  std::vector<unsigned int> _vals;
```
I don't think this should be a vector.  Rather than accumulating state, you just overwrite states progressively.

jthano replied:
I think these attrib classes are somewhat confusing, including things like this data member being a vector. But I would prefer to leave `AttribPreAux` as similar to `AttribExecOns` as possible because they now behave similarly.

As noted in a separate comment, the only difference between the two classes now is how the _vals object is filled.

jthano replied:
I still want `AttribPreAux` and `AttribPostAux` to be similar to `AttribExecOns`, but if you have a strong opinion about this data structure, I am will to change the vector to a set

rwcarlsen commented on framework/include/loops/ComputeUserObjectsThread.h:
```diff
@@ -118,10 +135,16 @@ groupUserObjects(TheWarehouse & w,
+        ic_deps.count(obj->name()) > 0)
     {
-      w.update(obj, AttribPreAux(w, true));
+      pre_aux_dependencies[obj].insert(Moose::ALL);
```
The Moose::ALL isn't meant to refer to execute_on flags - this makes the code confusing here.

jthano replied:
I was copying the functionality of the `AttribExecOns::isMatch(const Attribute & other) const` function where the flags are checked against `Moose::ALL` like so:
```c++
bool
AttribExecOns::isMatch(const Attribute & other) const
{
  auto a = dynamic_cast<const AttribExecOns *>(&other);
  if (!a || a->_vals.size() < 1)
    return false;
  auto cond = a->_vals[0];
  if (cond == Moose::ALL)
    return true;

  for (auto val : _vals)
    if (val == Moose::ALL || val == cond)
      return true;
  return false;
}
```
If this implementation will not work for `AttribPreAux` I can change it. Is there some other flag that indicates execute always?

jthano replied:
Also, I want to note that I removed the condition `aux_deps.count(obj->name()) > 0`. Maybe this bit of code is causing some confusion, it does not force the UO into `PRE_AUX` anytime there is a dependency. That is what the old code did, not the code in this MR.

The UO is still forced into PR_AUX if `ic_deps.count(obj->name()) > 0`. **Maybe this should be changed? Does it make sense to only force the UO into the EXEC_INITIAL group if an IC depends on it?**

jthano replied:
I need to think more about the UO grouping in this MR. As the code is currently in this MR, a UO can end up in more than one group, but that did not happen in the old code. I'll push a code update that fixes this

jthano replied:
I did change this in the most recent code update. I no longer use `MOOSE::ALL` for anything in `AttribPreAux` or `AttribExecOns`

rwcarlsen reviewed (changes requested):
It seems to me like this doesn't quite do what it sets out to.  What you really want is a particular UO to run either pre aux, pre ic, post aux depending on what objects depend on it for the *current* execute on flag.  But this code promotes a UO to pre aux for _all_ execute on flags if an aux kernel running on only _one_ execute on flag depends on that uo.  Is this what you wanted?  Or do we need to revisit the implementation to try to achieve the full handling I just described?

jthano commented:
> But this code promotes a UO to pre aux for _all_ execute on flags if an aux kernel running on only _one_ execute on flag depends on that uo. Is this what you wanted? Or do we need to revisit the implementation to try to achieve the full handling I just described?

I don't think this is the case, the test I added I think shows this does not occur. In the test I added, the section of input shown below is the important part. Before the code changes proposed in this MR, the `scale1` UO is forced into the `PRE_AUX` group because `NormalizationAuxTest1` depends on `scale1`. But `NormalizationAuxTest1` only execs on `FINAL` so there is not a dependency for any other exec flag. When I run this test with the code in this MR, I see that `scale1` is not run in `PRE_AUX` except for `FINAL`

```
[AuxKernels]
  active = 'NormalizationAuxTest1'
  [NormalizationAuxTest1]
    type = NormalizationAux
    variable = w
    source_variable = u
    normalization = scale1
    execute_on = 'FINAL'
  []
  ...
  [Postprocessors]
  ...
  [./scale1]
    type = ScalePostprocessor
    value = total_u
    scaling_factor = 1
  [../]
```

jthano commented:
@rwcarlsen Let me think more about the interaction between `PRE_AUX` and `POST_AUX`. I think in the existing code, a UO would always only be in either `PRE_AUX` or `POST_AUX`, but in my update, a UO could be in `PRE_AUX` and `POST_AUX`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/774945/) on e5d36da wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18293/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18293/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format eecd35ba285d56717f56887983cf29bf76ebfbcf`

rwcarlsen commented:
Let me know when you are ready for me to take another look.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/777900/) on 7cee761 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/18293/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/18293/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a969d327c03b526dd25c1af5b93f3b3aeabe2814`

Commit:
Addressing comments about documentation for mr #18293

jthano commented:
@rwcarlsen This is ready for a second round of review. The major update to the code I made since the first round is that `AttribPostAux` was changed to take an exec flag like `AttribPreAux`. Now, each exec flag is added either to `AttribPostAux` or `AttribPreAux` for a UO. I'll note that the code in this MR does fix an issue we have in Griffin which motivated the original issue.

YaqiWang commented:
@rwcarlsen can you review this again? @jthano you can do a rebase and push again. I think now next branch is in a good state and we should not see this many failures.

loganharbour commented:
There should be no need to rebase. All testing puts starts with devel as the base. I just invalidated the failing tests, which will trigger the new base off of devel.

The python failures were valid - there is an SQA issue.

Commit:
Updated test for PR #18293 to make coverage better and make compliant with sqa

rwcarlsen commented on framework/src/base/Attributes.C:
```diff
@@ -309,38 +309,68 @@ AttribPreIC::isEqual(const Attribute & other) const
+  if (!a || a->_vals.size() != _vals.size())
+    return false;
+
+  for (size_t i = 0; i < a->_vals.size(); i++)
+    if (a->_vals[i] != _vals[i])
+      return false;
```
Same here with the sorting.

jthano replied:
I made `_vals` for `AttribPreAux` and `AttribPostAux` a `std::set` instead of a `std::vector`. I think this change addresses your comments about the order of _vals. Now `isEqual()` just checks if the sets are `==`.

I had to make other changes to get this to work. I think the reason that `AttribExecOns` used a `std::vector` for `_vals` and compared the order in `AttribExecOns::isEqual()` is because of the member function `hashfunc(_vals)`. The hash function specialization for std::vector just takes the elements of `_vals` sequentially and returns a hash value based on that input. So vectors in different orders will return different hash values. I made an additional specialization for std::set so that now `AttribPreAux::isEqual()` and `AttribPostAux::isEqual()` will return true when the sets are the same and also `hashfunc(_vals)` will return the same hash value. You can see this small code change in MooseHashing.h

rwcarlsen reviewed (changes requested):
Looking pretty good.  Just a couple minor things.

Commit:
updated changed attribute objects to use std::set instead of std::vector to address comments

rwcarlsen commented on framework/src/base/Attributes.C:
```diff
@@ -309,38 +311,72 @@ AttribPreIC::isEqual(const Attribute & other) const
+  if (!a)
+    return false;
+
+  mooseAssert(a->_vals.size() == 1,
+              "Anticipated this function being passed an AttribPreAux object with a _vals member "
+              "containing only one item");
```
I'm not sure this assert is quite right.  I think it would be safer to assert that neither `this->_vals` nor `a->_vals` are size zero.  Since it is possible that either one could have more than one entry - we want to preserve symmetry - i.e. `foo->isMatch(bar)` should return the same result as `bar->isMatch(foo)`.

jthano replied:
Let me explain why I currently have `a->_vals` checking that there is exactly one element.  I think this is still the correct way to do this, but let me know what you think.

`FEProblemBase::execute` calls `FEProblemBase::computeUserObjects` which then calls `FEProblemBase::computeUserObjectsInternal` and the first part of this routine is shown below. So the idea is that when `AttribSubdomains::isMatch` is called, it is passed an Attib object which was created in the way shown below, that is with exactly one exec flag. 

```c++
void
FEProblemBase::computeUserObjectsInternal(const ExecFlagType & type,
                                          const Moose::AuxGroup & group,
                                          TheWarehouse::Query & query)
{
  if (group == Moose::PRE_IC)
    query.condition<AttribPreIC>(true);
  else if (group == Moose::PRE_AUX)
    query.condition<AttribPreAux>(type);
  else if (group == Moose::POST_AUX)
    query.condition<AttribPostAux>(type);
```

Although with the code as it currently is, `a->_vals` would always only have one element, someone could change the code in the future such that the query attrib object is passed a std::set of exec flags instead of just one. The reason I want to trip an assertion in this case is that I'm not sure what this future code developer wants from  `AttribSubdomains::isMatch` . For example, maybe the query attrib is passed a std::set and the code developer expects  `AttribSubdomains::isMatch` to return true if any one exec flag in the query object is in the attrib object for a UO. Or maybe they want  `AttribSubdomains::isMatch`  to return true if  the querry set is a subset of the attrib UO object. 

I also don't think the assertion should trip if `_vals` is empty. If a UO had an `AttribPreAux` or `AttribPostAux` with an empty `_vals` member, it would simply mean they don't execute preaux of postaux on any exec flag.

rwcarlsen replied:
Your assertion works fine for the current code as it stands in moose - but within the APIs intended use constraints, it is possible (or should be possible) to write code that would trip your assertion incorrectly.  The match and equality attributes need to operate symmetrically.  If `foo._vals.size()` is one and `bar._vals.size()` is 3, `foo.isMatch(bar)` needs to return the same as `bar.isMatch(foo)`.  The assertion as currently written would assert/fail in one case, but not the other.

jthano replied:
isMatch is now symmetric. As we discussed elsewhere, the assertion is removed

YaqiWang commented:
@GiudGiud Any progress on the BISON patch?

GiudGiud commented:
Some progress. Once #18593 goes in I will submit a patch to Bison. Then this PR will pass tests. 
@rwcarlsen reading all the conversations I do not get the feeling that this PR can currently be merged. Could you please make sure that are happy with all the requested modifications?

jthano commented:
@GiudGiud sorry I need to push one more small code update to address a final comment @rwcarlsen had made. I was waiting for your PR to be merged before making my final code update

Commit:
Merge remote-tracking branch 'upstream/next' into pre_aux_grouping_update

Commit:
updates to address final comments and update code after other changes made to the code base related to preaux postaux ordering

Commit:
removed GeneralUserObject documentation change since the parameter force_preaux and force_preic now apply to userobjects

GiudGiud commented:
@jthano could you please resolve all conversations then hmu I ll merge it

GiudGiud commented:
Thanks for the fix!
